overlord/configstate/configcore/handlers.go

```diff
@@ -0,0 +1,143 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/overlord/configstate/config"
+	"github.com/snapcore/snapd/release"
+)
+
+type configHandler interface {
+	validate(config.ConfGetter) error
+	handle(config.ConfGetter, *fsOnlyContext) error
+	needsState() bool
+	flags() flags
+}
+
+// flags carries extra flags that influence how the handler is called.
+type flags struct {
+	// coreOnlyConfig tells Run/FilesystemOnlyApply to apply the config on core systems only.
+	coreOnlyConfig bool
+}
+
+type fsOnlyHandler struct {
+	validateFunc func(config.ConfGetter) error
+	handleFunc   func(config.ConfGetter, *fsOnlyContext) error
+	configFlags  flags
+}
+
+var handlers []configHandler
+
+func init() {
+	// Most of these handlers are no-op on classic.
+	// TODO: consider allowing some of these on classic too?
+	// consider erroring on core-only options on classic?
+
+	coreOnly := &flags{coreOnlyConfig: true}
+
+	// watchdog.{runtime-timeout,shutdown-timeout}
+	addFSOnlyHandler(validateWatchdogOptions, handleWatchdogConfiguration, coreOnly)
+
+	// Export experimental.* flags to a place easily accessible from snapd helpers.
+	addFSOnlyHandler(validateExperimentalSettings, doExportExperimentalFlags, nil)
+
+	// network.disable-ipv6
+	addFSOnlyHandler(validateNetworkSettings, handleNetworkConfiguration, coreOnly)
+
+	// service.*.disable
+	addFSOnlyHandler(nil, handleServiceDisableConfiguration, coreOnly)
+
+	// system.power-key-action
+	addFSOnlyHandler(nil, handlePowerButtonConfiguration, coreOnly)
+
+	// pi-config.*
+	addFSOnlyHandler(nil, handlePiConfiguration, coreOnly)
+}
+
+// addFSOnlyHandler registers functions to validate and handle a subset of
+// system config options that do not require to manipulate state but only
+// the file system.
+func addFSOnlyHandler(validate func(config.ConfGetter) error, handle func(config.ConfGetter, *fsOnlyContext) error, flags *flags) {
+	h := &fsOnlyHandler{
+		validateFunc: validate,
+		handleFunc:   handle,
+	}
+	if flags != nil {
+		h.configFlags = *flags
+	}
+	handlers = append(handlers, h)
+}
+
+func (h *fsOnlyHandler) needsState() bool {
+	return false
+}
+
+func (h *fsOnlyHandler) flags() flags {
+	return h.configFlags
+}
+
+func (h *fsOnlyHandler) validate(cfg config.ConfGetter) error {
+	if h.validateFunc != nil {
+		return h.validateFunc(cfg)
+	}
+	return nil
+}
+
+func (h *fsOnlyHandler) handle(cfg config.ConfGetter, opts *fsOnlyContext) error {
+	if h.handleFunc != nil {
+		return h.handleFunc(cfg, opts)
+	}
+	return nil
```

Comment:
While this is okay are we expecting to see any non-handled cases? I'm worried that we will silently ignore something and be non-the-wiser until we trace it back to this `return nil`.

Reply:
Yes. Handle is implemented when we need to act on config values (e.g. when a core config entry needs to be reflected in the filesystem or elsewhere). There are cases where we only validate config values and they are used by other parts of the snapd - e.g. validateAutomaticSnapshotsExpiration, validateRefreshSchedule and validateRefreshRateLimit have no handlers.

Reply:
We can probably though add a few more checks, be a bit more explicit. I'll write some suggestions in a bit.

Reply:
Done.